src/states/elder/mod.rs

```diff
@@ -1706,6 +1658,48 @@ impl Approved for Elder {
 
         Ok(())
     }
+
+    // Checking members vote status and vote to remove those non-resposive nodes.
+    // Mock parsec will fetch all peers' votes in one poll loop, hence doesn't need the logic of
+    // expecting extra accusations. And even may have the risk of not having enough accusations if
+    // switch to the same code as for real parsec.
+    #[cfg(feature = "mock_parsec")]
+    fn check_vote_status(&mut self) {
+        let unresponsive_nodes = self.chain.check_vote_status();
+        for pub_id in unresponsive_nodes {
+            self.vote_for_event(AccumulatingEvent::Offline(pub_id));
+        }
+    }
+
+    #[cfg(not(feature = "mock_parsec"))]
+    fn check_vote_status(&mut self) {
+        let unresponsive_nodes = self.chain.check_vote_status();
+        for (id, count) in mem::replace(&mut self.unresponsive_nodes, Default::default()) {
+            if unresponsive_nodes.contains(&id) {
+                let new_count = count.saturating_add(1);
```

Comment:
So I can see how that works with some bulk group of vote happening, some nodes are behind and detected as unresponsive, and as time of, all there votes come in and they are no longer unresponsive.

What is not clear to me is how that work with continuous voting happening.
The first time, a slower node votes may not have been there yet, counted as unresponsive
The second time, their vote for the first batch has come, but the window as moved, and their vote for the second batch are not there yet, still counted as unresponsive.
The third time, window continue to move, and so there older votes coming still not make them look responsive.
etc.. eventually they are voted offline for the reason they are a bit behind (not much just enough for a sufficient length of time).

I'm thinking we can write a test for this that would fail for that code, and also represent more accurately the behaviour of the network (i.e continuous events voted).

Reply:
> eventually they are voted offline for the reason they are a bit behind (not much just enough for a sufficient length of time).

It means it needs to be lagged for THRESHOLD number of consensused events behind. Which shall not happend, as its vote shall come in within Section_Member_Size parsec polls normally (same reasoning as at https://github.com/maidsafe/routing/pull/1880#discussion_r349081366 )

> I'm thinking we can write a test for this that would fail for that code

Not sure about this. The test sounds more likely that node to be `muted` actually  :(

Reply:
I do not think that is reasonable to say the node shall not lag by `UNRESPONSIVE_THRESHOLD` or even `UNRESPONSIVE_WINDOW` would not happen:

 example:

```
Network: 200 events per second require voting by all nodes.
All voters:      Vote 200 events -1sec> Vote 200 events -1sec> Vote 200 events -1sec> Vote 200 events -1sec> ...
Lagging by 2 second:     Nothing -1sec> Nothing         -1sec> Vote 200 events -1sec> Vote 200 events -1sec> ...
Consensus:
 Nothing
 -1sec> 200 events consensused for all voters, but lagging node: Lagging node unresponsive.
 -1sec> 200 events consensused for all voters, but lagging node: Lagging node unresponsive.
 -1sec> 200 events consensused for all voters, lagging node votes for first 200 events, all these are no longer in the window: Lagging node unresponsive.
 -1sec> 200 events consensused for all voters, lagging node votes for second 200 events, all these are no longer in the window: Lagging node unresponsive.
 -1sec> 200 events consensused for all voters, lagging node votes for third 200 events, all these are no longer in the window: Lagging node unresponsive.
...
 -> Lagging node voted offline.
```

Reply:
I think if we expect 200 (or whatever) events per one gossip, then we would need to adjust the windows and threshold accordingly. Also, if a node lags that much then perhaps it's OK to consider it as unresponsive?

Reply:
Alternatively, the window can be measured not in events but something else. Number of gossips? Or even actual time (seconds)? That could come with its own set of issues though...

Reply:
Well not handling lagging node may be a problem here if we want to use that mechanism to support upgrade where the node could possibly drop out for a bit, so is 2 second lagging too much?

Moreover, that 2 second is with the gossip event every second, something we would likely want to decrease to decrease parsec latency, as a result a much smaller lag may be enough. 

Maybe a fixed window is problematic if what we care about is that they voted for some event.
If what we are looking for is did they vote for enough event during the multiple gossip event we got, then maybe that window should reflect that.

But there are some things for consideration:
 - I am not clear that the way we are calling `self.check_vote_status();` make it shared state (i.e I think different nodes will call it after different event, I think it should be shared state if called from within the `while let Some(block) = self.parsec_map_mut().poll() {` block, I think, but then it is also not related to a gossip event being received).
 - (There may be consideration of malicious attacks by generating gossip, but that does not seem to be the case).
 - (a secondary consideration would be a window growing too big be problematic in term of performance.)


I'm sorry I'm not really proposing solutions here, I would like to, but I have yet to see a way that seem consistent with a flow of events being consensused. I think what you are saying @madadam may be correct in that maybe the window should be adjusted. But if the window is sufficient so we can handle these, then what is the purpose of the second windowing happening where we validate over multiple gossip event?

Reply:
> Maybe a fixed window is problematic

could be. However to make all members using the same window size is another shared_state need to be exchanged. :(

>  called from within the `while let Some(block) = self.parsec_map_mut().poll() {` block

Previously it was. However it incur a performance regression. Maybe what we can do is: remain at the present calling place but with an extra condition that there shall be something polled from parsec this time.

> But if the window is sufficient so we can handle these, then what is the purpose of the second windowing happening where we validate over multiple gossip event

This is mainly due to the testing pattern: number of injected user-data depends on the window size. unless we change it to what Adam suggested that injecting user-data gradually.

Reply:
> Maybe what we can do is: remain at the present calling place but with an extra condition that there shall be something polled from parsec this time.

I am not clear how that solve the problem highlighted.

Reply:
Same as to calling inside I think. just reduce multiple calls into one time call.

Reply:
With gossip
```
A10 <- B9
 |      |    
 |     B8
 |      |
A9 ->- B7 
``
B may see a block for A9, loop exit, a block for B7, loop exit, block for B8, loop exit, Block for B9, loop exit.
A may see a block for A9, loop exit, a block for B7, then B8, then B9, then A9, then A10 loop exit.

The reason is when getting gossip creating A10, Alice got all the other sync event that B saw one at a time.